File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def recursive(self,arr, rem, index):
        if rem < 0 or index < 0: 
            return False
        if rem == 0:
            return True
        take = self.recursive(arr, rem - arr[index], index - 1)
        not_take = self.recursive(arr, rem, index - 1)
        return take or not_take
    
    def memoized(self,arr, rem, index, dp):
        if rem == 0:
            return True
        if rem < 0 or index < 0: 
            return False
        if dp[index][rem] != -1:
            return dp[index][rem]
        take = self.memoized(arr, rem - arr[index], index - 1, dp)
        not_take = self.memoized(arr, rem, index - 1, dp)
        dp[index][rem] = take or not_take
        return dp[index][rem]
    
    def tabulation(self,arr, N, sum):
        dp = [[False for _ in range(sum + 1)] for _ in range(N)]
        for index in range(N):
            dp[index][0] = True
        for index in range(N):
            for rem in range(sum + 1):
                if rem == 0:
                    dp[index][rem] = True
                else:
                    take = False
                    not_take = False
                    if rem - arr[index] >= 0:
                        take = dp[index - 1][rem - arr[index]]
                    not_take = dp[index - 1][rem]
                    dp[index][rem] = take or not_take
        return dp[N - 1][sum]
    
    def optimized_tabulation(self, arr, N, sum):
        prev = [False for _ in range(sum + 1)] 
        curr = [False for _ in range(sum + 1)] 
        for index in range(N):
            for rem in range(sum + 1):
                if rem == 0:
                    curr[rem] = True
                else:
                    take = False
                    if rem - arr[index] >= 0:
                        remainder = rem - arr[index]
                        if remainder == 0 and index == 0:
                            take = True
                        else:
                            take = prev[rem - arr[index]]
                    not_take = prev[rem]
                    curr[rem] = take or not_take
            prev, curr = curr, prev
        return prev[sum]

    def canPartition(self, nums: List[int]) -> bool:
        n = len(nums)
        total = sum(nums)
        if total % 2 == 1:
            return False
        target = total // 2
        return self.optimized_tabulation(nums, n - 1, target)

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    // // TC -> O(N * 2 ^ N);
    // // SC -> O(N)
    // public boolean recursion(int[] nums, int index, int remainder){
    //     if (remainder < 0){
    //         return false;
    //     }
    //     if (index == 0){
    //         if (nums[index] == remainder){
    //             return true;
    //         }else{
    //             return false;
    //         }
    //     }

    //     boolean take = recursion(nums,index - 1, remainder - nums[index]);
    //     boolean not_take = recursion(nums, index - 1, remainder);

    //     return take || not_take;
    // }

    public boolean memoized(int[] nums, Boolean[][] dp, int index, int remainder){
        if (remainder < 0){
            return false;
        }
        if (index == 0){
            if (nums[index] == remainder){
                return true;
            }else{
                return false;
            }
        }
        if(dp[index][remainder] != null){
            return dp[index][remainder];
        }

        boolean take = memoized(nums, dp, index - 1, remainder - nums[index]);
        boolean not_take = memoized(nums,dp, index - 1, remainder);

        dp[index][remainder] = take || not_take;
        return dp[index][remainder];
    }

    public boolean canPartition(int[] nums) {
        int total = 0;
        for(int num : nums){
            total += num;
        }
        if (total % 2 == 1){
            return false;
        }
        int arr_size = nums.length;
        Boolean[][] dp = new Boolean[arr_size][total];
        for (int r = 0; r < arr_size; r++){
            for (int c = 0; c < total; c++){
                dp[r][c] = null;
            }
        }
        return memoized(nums, dp, nums.length - 1, total / 2);
    }
}